### PR TITLE
Make process RAM Read-Write, not Read-Write-Execute

### LINKED ISSUE #1287: Change default application RAM permissions to RW (remove execute permission)
We should implement write-xor-execute protections by default, to defend against remote code execution attacks on applications. This can be done by changing the default MPU permissions for application memory from RWX to RW.

I'll open a separate issue to define and implement a mechanism for applications to mark RAM as executable.

### PR DESCRIPTION
**Pull request by alevy**
### Pull Request Overview

Fixes #1287 

Removes execute permission for process RAM to help defend against code execution attacks on applications.

### Testing Strategy

First, I ran a number of apps (`ip_sense`, `c_hello`, `blink`), to verify sure this doesn't break anything in unexpected ways (the changes should have no impact on existing apps, since nothing runs from RAM in practice).

Next, I wrote a simple test app that includes a RAM allocated variable containing _just_ an ARM `bx lr` instruction (`0x4770`):

```
#include <stdio.h>

#include <console.h>
#include <timer.h>

uint16_t space = 0x4770;

typedef void dummy(void);

int main(void) {
  int j = (int)(&space) | 1;
  asm ("blx %0;" : : "r"(j));
  while (1) {
    printf("Hello\n");
    delay_ms(1000);
  }
}
```

Running this on `master` works: the app doesn't crash, and it prints `Hello` periodically following a brief prelude in the RAM-allocated "dummy" function.

Running on this branch, with write permissions turned off for RAM, the app crashes with an access violation fault, as expected:

```
Kernel panic at /home/alevy/hack/tock/kernel/src/process.rs:537:
        "Process rwx had a fault"
        Kernel version release-1.2-2018-06-926-g9637f419

---| Debug buffer not empty. Flushing. May repeat some of last message(s):
Initialization complete. Entering main loop

---| Fault Status |---
Instruction Access Violation:       true
Forced Hard Fault:                  true
Fault Status Register (CFSR):       0x00000001
Hard Fault Status Register (HFSR):  0x40000000

...
```

### TODO or Help Wanted

Where should we document this?

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: kernel/src/process.rs [hunk 1]
```diff
@@ -704,7 +704,7 @@ impl<C: Chip> ProcessType for Process<'a, C> {
                 unallocated_memory_start,
                 unallocated_memory_size,
                 min_region_size,
-                mpu::Permissions::ReadWriteExecute,
+                mpu::Permissions::ReadWriteOnly,
```
Comment (rajivr) [inline kernel/src/process.rs]: I think we use `allocate_region` is used to allocate MPU region for flash memory. Shouldn't this be `mpu::ReadExecuteOnly`?

Comment (alevy) [reply kernel/src/process.rs]: `add_mpu_region` is used only for IPC regions. The flash region is allocated on line 1163 as `ReadExecuteOnly` (which hasn't changed).

Comment (rajivr) [reply kernel/src/process.rs]: That's right. Sorry, I forgot about IPC use-case. Thanks for clarifying.

**Comment by alevy** [comment]
This is ready for review. I think the changes are fairly straight forward and simple, but don't let me cloud your judgement

**Pull request merged**